packages/astro/src/core/ssr/index.ts

```diff
@@ -274,6 +269,11 @@ export async function render(renderers: Renderer[], mod: ComponentInstance, ssrO
     html = await viteServer.transformIndexHtml(viteifyURL(filePath), html, pathname);
   }
 
+  // inject <!doctype html> if missing (TODO: is a more robust check needed for comments, etc.?)
+  if (!/<!doctype html/i.test(html)) {
+    html = '<!DOCTYPE html>\n' + html;
```

Comment:
`<!DOCTYPE` insertion saved until the very end, because we may be injecting things to the beginning of the document for partials.